cupy/indexing/generate.py

```diff
@@ -244,7 +248,105 @@ def ix_(*args):
+        dims (tuple of ints): The shape of array into which the indices from
+            ``multi_index`` apply.
+        mode ('raise', 'wrap' or 'clip'), optional: Specifies how out-of-bounds
+            indices are handled.  Can specify either one mode or a tuple of
+            modes, one mode per index.
+            * 'raise' -- raise an error (default)
+            * 'wrap' -- wrap around
+            * 'clip' -- clip to the range
+            In 'clip' mode, a negative index which would normally wrap will
+            clip to 0 instead.
+        order ('C' or 'F'), optional: Determines whether the multi-index should
+            be viewed as indexing in row-major (C-style) or column-major
+            (Fortran-style) order.
+
+    Returns:
+        raveled_indices (cupy.ndarray): An array of indices into the flattened
+            version of an array of dimensions ``dims``.
+
+    Examples
+    --------
+    >>> cupy.ravel_multi_index(cupy.asarray([[3,6,6],[4,5,1]]), (7,6))
+    array([22, 41, 37])
+    >>> cupy.ravel_multi_index(cupy.asarray([[3,6,6],[4,5,1]]), (7,6),
+    ...                        order='F')
+    array([31, 41, 13])
+    >>> cupy.ravel_multi_index(cupy.asarray([[3,6,6],[4,5,1]]), (4,6),
+    ...                        mode='clip')
+    array([22, 23, 19])
+    >>> cupy.ravel_multi_index(cupy.asarray([[3,6,6],[4,5,1]]), (4,4),
+    ...                        mode=('clip', 'wrap'))
+    array([12, 13, 13])
+    >>> cupy.ravel_multi_index(cupy.asarray((3,1,4,1)), (6,7,8,9))
+    array(1621)
+
+    .. seealso:: :func:`numpy.ravel_multi_index`, :func:`unravel_index`
+    """
+
+    ndim = len(dims)
+    if len(multi_index) != ndim:
+        raise ValueError(
+            "parameter multi_index must be a sequence of "
+            "length {}".format(ndim))
+
+    for arr in multi_index:
+        if not isinstance(arr, cupy.ndarray):
+            raise TypeError("elements of multi_index must be cupy arrays")
+        if arr.dtype.kind not in 'iub':
+            raise TypeError("only int indices permitted")
+
+    for d in dims:
+        if not isinstance(d, numbers.Integral):
+            raise TypeError(
+                "{} object cannot be interpreted as an integer".format(
+                    type(d)))
+
+    if isinstance(mode, str):
+        mode = (mode, ) * ndim
+
+    if functools.reduce(operator.mul, dims) > cupy.iinfo(cupy.int64).max:
+        raise ValueError("invalid dims: array size defined by dims is larger "
+                         "than the maximum possible size")
+
+    s = 1
+    ravel_strides = [1] * ndim
+    if order is None:
+        order = "C"
+    if order == "C":
+        for i in range(ndim - 2, -1, -1):
+            s = s * dims[i + 1]
+            ravel_strides[i] = s
+    elif order == "F":
+        for i in range(1, ndim):
+            s = s * dims[i - 1]
+            ravel_strides[i] = s
+    else:
+        raise TypeError("order not understood")
+
+    raveled_indices = 0
```

Comment:
Instead of setting it to 0, I feel its better to just allocate a `cupy.array` here rather than changing the type of this variable later inside the loop.